src/dataset.cpp

```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+                         : d(Coord::DetectorGrouping{}).dimensions();
+  Variable tofShift(Data::Value{}, {});
+  Variable scale(Data::Value{}, {});
+
+  if (d.contains(Coord::Ei{})) {
+    // Direct-inelastic.
+
+    // This is how we support multi-Ei data!
+    tofShift.setDimensions(d(Coord::Ei{}).dimensions());
+    const auto &Ei = d.get<const Coord::Ei>();
+    std::transform(Ei.begin(), Ei.end(), tofShift.get<Data::Value>().begin(),
+                   [&](const double Ei) { return l1 / sqrt(Ei); });
+
+    scale.setDimensions(dims);
+    MDZipView<const Coord::Position> specPos(d);
+    std::transform(specPos.begin(), specPos.end(),
+                   scale.get<Data::Value>().begin(), [&](const auto &item) {
+                     const auto &pos = item.template get<Coord::Position>();
+                     const double l2 = (samplePos - pos).norm();
+                     return l2 * l2;
+                   });
+  } else if (d.contains(Coord::Ef{})) {
+    // Indirect-inelastic.
+
+    tofShift.setDimensions(dims);
+    // Ef can be different for every spectrum so we access it also via a view.
+    MDZipView<const Coord::Position, const Coord::Ef> geometry(d);
+    std::transform(geometry.begin(), geometry.end(),
+                   tofShift.get<Data::Value>().begin(), [&](const auto &item) {
+                     const auto &pos = item.template get<Coord::Position>();
+                     const auto &Ef = item.template get<Coord::Ef>();
+                     const double l2 = (samplePos - pos).norm();
+                     return l2 * l2 / sqrt(Ef);
+                   });
+
+    scale.setDimensions({});
+    scale.get<Data::Value>()[0] = l1 * l1;
+  } else {
+    throw std::runtime_error("Dataset contains neither Coord::Ei nor "
+                             "Coord::Ef, this does not look like "
+                             "inelastic-scattering data.");
+  }
+
+  // 2. Transform variables
+  Dataset converted;
+  for (const auto &var : d) {
+    auto varDims = var.dimensions();
+    if (varDims.contains(Dim::Tof))
+      varDims.relabel(varDims.index(Dim::Tof), Dim::DeltaE);
+    if (var.tag() == Coord::Tof{}) {
+      auto dims = scale.dimensions();
+      for (const Dim dim : varDims.labels())
+        if (!dims.contains(dim))
+          dims.addInner(dim, varDims[dim]);
+      Variable E(Coord::DeltaE{}, dims, dims.volume(), 1.0);
+      E *= var.reshape(varDims);
+      E -= tofShift;
+      E *= E;
+      E = 1.0 / std::move(E);
+      E *= scale;
+      if (d.contains(Coord::Ei{})) {
+        converted.insert(-(std::move(E) - d(Coord::Ei{})));
+      } else {
+        converted.insert(std::move(E) - d(Coord::Ef{}));
+      }
+    } else if (var.tag() == Data::Events{}) {
+      throw std::runtime_error(
+          "TODO Converting units of event data not implemented yet.");
+    } else {
+      converted.insert(var.reshape(varDims));
+    }
+  }
+
+  return converted;
+}
```

Comment:
Given the flexibility of what can be done here, in this units like approach the implementation of TOF -> Q3D (histogram) would look very similar.  Potentially, `Q1D`, `Qxy`, `SofQW`, `ConvertToReflectometryQ` .... could also be done in this form.

Reply:
Yes, see the long comment in the body of the `convert` function. The main question is whether the conversion is uniquely specified by providing dimensions?

For example:

```cpp
auto q1d = convert(dataset, {Dim::Tof, Dim::Spectrum}, {Dim::Q});
auto qxy = convert(dataset, {Dim::Tof, Dim::Spectrum}, {Dim::Qx, Dim::Qy});
```

Is this completely unambiguous?

Reply:
Given your `Ei`, `Ef` improvements, I think it's definitely easier to deduce what is possible, though for cases such as elastic conversions, you may also (optionally) convert into different frames (QLab, QSample, etc).

Reply:
Would that just be a different dimension, `Dim::QLabel` and `Dim::QSample`?